pkg/providers/cloudstack/validator.go

```diff
@@ -130,13 +130,21 @@ func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, clusterSp
 		if !ok {
 			return fmt.Errorf("cannot find CloudStackMachineConfig %v for worker nodes", workerNodeGroupConfiguration.MachineGroupRef.Name)
 		}
+		// validate template field of worker group spec with the kubernetes version of each workerNodeGroup - in case of modular upgrade.
```

Comment:
I don't think the algorithm here is correct. We need to validate the following:

1. The control plane's machine config template specifies the top level Kubernetes version. 
2. For each worker node group, if the worker node group specifies its own Kube version, that version exists in the machine configs template, _else_ the top level Kube version eixsts in the machine configs template.

L142-150 are checking each machine config has the root Kube version, but that's not necessarily what we expect due to modular upgrade (point 2 above).

The way this algorithm is structured makes the validation challenging to follow and the resulting error messages are unnecessarily long and contain detail that ought to live on the API documentation. Given we're doing some work in this area can we take a moment to refactor that also?

See https://github.com/aws/eks-anywhere/pull/6805/files#diff-d794b88ef59dfa162a093bf0492363f2f03fb35c96e442ca1644ef4dea859b5eR76-R97 which contains the Tinkerbell check that's fundamentally the same.

Reply:
Makes sense. I have reimplemented the algorithm. It satisfies both 1 and 2 points. Regarding the error message, since this is a new ask from the user, we wanted to make it more obvious to understand the behavior here. We discussed this in this PR comment [here](https://github.com/aws/eks-anywhere/pull/6728#discussion_r1336417088) and [here](https://github.com/aws/eks-anywhere/pull/6728#discussion_r1337799633). I am open to any other suggestions.

Reply:
I empathize with the concern and what we're trying to do, but I don't agree with the approach. There are a few reasons we shouldn't include all the current detail: 

1. Errors should concisely state the problem and leave describing valid inputs and examples to documentation. 
2. It can be challenging to adequately describe valid input for all inputs. For example, perhaps some inputs require a more elaborate explanation and examples that can't be added to errors. We start generating inconsistent errors when we offer examples and explanations in some cases but not others.
3. Documentation and code fall out of sync (I [raised a bug](https://github.com/aws/eks-anywhere/issues/6836) to that effect recently). If we change expectations as we have on a handful of occasions we need to regurgitate the API documentation in this error message; its a maintenance overhead.

If we want to provide remedial recommendations in code I think we need a more comprehensive approach.